### ISSUE TITLE
Charts to be responsive on screen resize

### ISSUE THREAD

**Issue by richavyas** [swimlane/ngx-charts#147]
**I'm submitting a 
- [X] feature request

**Current behavior**
Charts are not responsive. On screen resize, they don't scale.

**Expected behavior**
Charts should resize based on screen resize.

**Reproduction of the problem**
![swimlane-charts-responsive](https://cloud.githubusercontent.com/assets/4469548/21580184/f8e0d610-cf96-11e6-85ea-49ee15b4fd65.gif)

**What is the motivation / use case for changing the behavior?**
To use charts with different screen sizes

**Please tell us about your environment:**
Operating system - mac
IDE - webstorm
package manager - NPM

* **ngx-charts version:** 3.0.4

* **Angular version:** 2.4.1

* **Browser:**Chrome & Firefox

**Comment by marjan-georgiev** [comment]
Try removing the `[view]` input. If that is not present, the chart resizes to fit its parent container.

**Comment by kyleledbetter** [comment]
we can prob close this one in favor of this feature request https://github.com/swimlane/ngx-charts/issues/150

**Comment by Asone** [comment]
@marjan-georgiev : if view is undefined the charts fits to its parent container, true, but on resize, if the parent container is responsive ( c.f bootstrap class container ) the chart doesn't resize. 

:(

**Comment by Asone** [comment]
@marjan-georgiev : tried it on v5.3.1

**Comment by markmssd** [comment]
I agree with @Asone, is there any way to force a resize when the parent container is resized?

**Comment by williangd** [comment]
Same problem described by @Asone here. @markmssd, Have you found any workarounds?

**Comment by markmssd** [comment]
Nope no solution yet, just living with it for now...

**Comment by dudipsh** [comment]
Hey 
you can try to bind the view array when window resize
 work for me

component
`
onResize(event) {
    this.view = [event.target.innerWidth - 900, 280 ];
  }
`
in html 
`
<ngx-charts-area-chart  (window:resize)="onResize($event)"></ngx-charts-area-chart>
`

**Comment by markmssd** [comment]
Where are you getting the `view` array from?

**Comment by mmahmad** [comment]
@dudipsh that works until a certain point, after which it throws a negative width error
![image](https://user-images.githubusercontent.com/4930049/33163363-06d31f02-d050-11e7-86b1-a2ddced9e935.png)

**Comment by dudipsh** [comment]
@mmahmad 
It seems that the size you set is negative
In onResize console log(event.target.innerWidth - 900)

**Comment by Sendrea** [comment]
Put chart in a container

**Comment by caprushit1** [comment]
Put chart in a container ,even if chart not resizing properly , just fire the window resize event it will automatically resize.  **window.dispatchEvent(new Event('resize'));**

**Comment by dudipsh** [comment]
`  
  @ViewChild('graphWrapper') graphWrapper: ElementRef;

fixGraphPosition() {
    this.zone.runOutsideAngular(() => {
      Observable.fromEvent(window, 'resize')
        .debounceTime(200).distinctUntilChanged().subscribe((e: Event) => {
          this.zone.run(() => {
            this.changeSize.next(e);
          })
        }
        )
    });
    this.changeSize.subscribe((e: Event) => { this.onResize(e); });
 onResize(event) {
    if (this.graphWrapper && this.graphWrapper.nativeElement) {
      this.chartWidth = this.graphWrapper.nativeElement.offsetWidth;
    }
  }
`

**Comment by rkuodys** [comment]
@mmahmad Sorry for off topic question, but how did you get this nice donut chart? I've spent couple of hours so far struggling with this to take out of the box. By any chance there is a manual I could use? Thx in advance.

**Comment by noeldelgadom** [comment]
I solved it by putting it inside a div with display: grid;

// html
<div id="pie-chart-parent">
    <ngx-charts-pie-chart></ngx-charts-pie-chart>
</div>

// css
#pie-chart-parent {
    display: grid;
}

**Comment by noeldelgadom** [comment]
I solved it by putting it inside a div with display: grid;

// html

    <div id="pie-chart-parent">

        <ngx-charts-pie-chart></ngx-charts-pie-chart>

    </div>


// css
`#pie-chart-parent {
display: grid;
}`

**Comment by torontocode** [comment]
I went a bit old school with this. After spending ages on it, I solved it by adding a 1px transparent .gif into the responsive container which houses, say, `ngx-charts-pie-chart` :

`<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7">`

then style it:

`img {
  width: 100%;
}
ngx-charts-pie-chart{
  position:absolute;
  top:0;
  bottom:0;
  left:0;
  right:0;
}`

This works because the image forces a percentage height which the chart needs to size itself. 

It is slow to resize in that there is a slight delay after the browser window has been resized, but it does work if you don't mind that.

**Comment by niddynod** [comment]
hi @torontocode, do you still use the img tag for new solution? or just a container class you create that is the parent to the chart

**Comment by torontocode** [comment]
No - you don't need an image any more - just set chart css `position` to `absolute` and `padding-bottom:100%` on the parent.

**Comment by ErvinLlojku** [comment]
I had the same problem with treemap chart. My solution:
``` HTML
<ngx-charts-tree-map
  #chart
  [view]="[width, height]"
  [scheme]="colorScheme"
  [results]="single"
  [valueFormatting]="valueFormatting">
</ngx-charts-tree-map>
```
``` Typescript
@Input() private _width = null;
  set width(value: any) {
    this._width = value;
  }
  get width(): any {
    return this._width || this.el.nativeElement.clientWidth;
  }
constructor(private el: ElementRef) {}
```
Logic: If no width specified than i take the width of the host element. 
``` Css
:host {
  display: block;
}
```

**Comment by amrit-kumar** [comment]
@dudipsh My onResize function does not get called .I am using [https://github.com/tiberiuzuld/angular-gridster2](angular-gridster2)  with ngx-charts so I need to resize the chart based on div size .Here is how I have used   ```
<div #component1 class="ngx-charts-bar-vertical" style=" display: grid; ">
  <ngx-charts-bar-vertical
  (window:resize)="onResize($event)" 
  [scheme]="colorScheme"
  [results]="single"
  [gradient]="gradient"
  [xAxis]="showXAxis"
  [yAxis]="showYAxis"
  [legend]="showLegend"
  [showXAxisLabel]="showXAxisLabel"
  [showYAxisLabel]="showYAxisLabel"
  [xAxisLabel]="xAxisLabel"
  [yAxisLabel]="yAxisLabel"
  [xAxisTickFormatting]="xAxisTickFormatting"
  (select)="onSelect($event)"
  (legendLabelClick)="onLegendLabelClick($event)"
  (dblclick)="dblclick($event)"
  (xAxisLabelclick)="xAxisLabelclick($event)">
  
  </ngx-charts-bar-vertical>
</div>
```

Any help on how can i resize the chart based on user interaction with div .

**Comment by dudipsh** [comment]
@amrit-kumar 
did you try with my example? (its works for me)
try to move the " (window:resize)="onResize($event)" up to the div
`<div #component1 class="ngx-charts-bar-vertical" (window:resize)="onResize($event) ">
<ngx-charts-bar-vertical></<ngx-charts-bar-vertical>
</div>`

**Comment by wahabiberexia** [comment]
create 3 charts with difference size :dagger:

**Comment by khalidcawl** [comment]
How the chart width and height is determined is explained [here](https://github.com/swimlane/ngx-charts/issues/150#issuecomment-303129628).

So what I did is basically set the min-height of the chart's parent. The responsiveness upon window resize is taken care of automatically. 

```
<div class="row">
  <div class="col-md-6 chart-parent">
    <ngx-charts-bar-vertical [view]="view" ></ngx-charts-bar-vertical>
  </div>
  <div class="col-md-6 chart-parent">
     <ngx-charts-bar-vertical [view]="view" ></ngx-charts-bar-vertical>
  </div>
</div>

.chart-parent{
     min-height: 350px;
}
```
Both charts now resize nicely. 

Hope this helps someone.

**Comment by kuncevic** [comment]
Just found some solution here https://stackoverflow.com/questions/52552815/how-can-i-make-my-ngx-line-chart-responsive

**Comment by EfkanKnz** [comment]
The solution for me was to create a parent div and add some CSS :
`charts.component.html`
```
<div id="line-chart-parent" class="col-8">
 <ngx-charts-line-chart class="chart-container"
  [scheme]="colorScheme"
  [results]="results"
  [curve]="curve"
  [showXAxisLabel]="true"
  [showYAxisLabel]="true"
  [xAxis]="true"
  [yAxis]="true"
  [autoScale]="true"
  [showGridLines]="true"
  (select)="select($event)">
 </ngx-charts-line-chart>
</div>
```
`charts.component.css`
```
#line-chart-parent {
 display: grid;
 min-height: 500px;
}
```
I wanted the width to be responsive and the heigth to be fixed.

**Comment by yogendra-waghale** [comment]
@amit11patel6 you are the man!
We were trying it from last 3 days , but finally from your solution , we have done it!!!! Thanks a lot.

**Comment by rodrigosanchezg8** [comment]
This should be definitively handled by the ngx charts lib. Not setting view didn't do the trick, neither did [view]="". 

Here's my temporary "fix":

html
```
 <div #pbgChartContainer>
<ngx-charts-bar-horizontal
...
(window:resize)="onResize()"
...
</ngx-charts-bar-horizontal>
</div>
```

component
```
@ViewChild('pbgChartContainer') pbgChartContainer: ElementRef;
...
    onResize() {
        setTimeout(() => {
            this.pbgChartOptions.view = [this.pbgChartContainer.nativeElement.offsetWidth, 300];
        }, 500)
    }
```

**Comment by kavya-tatvasoft** [comment]
> Put chart in a container ,even if chart not resizing properly , just fire the window resize event it will automatically resize. **window.dispatchEvent(new Event('resize'));**

Hi can you please tell me exactly where I have to put this in my TS file?

**Comment by kavyahalani** [comment]
> Put chart in a container ,even if chart not resizing properly , just fire the window resize event it will automatically resize. **window.dispatchEvent(new Event('resize'));**

Hi can you please tell me exactly where I have to put this in my TS file?

**Comment by omdhanwant** [comment]
> > Put chart in a container ,even if chart not resizing properly , just fire the window resize event it will automatically resize. **window.dispatchEvent(new Event('resize'));**
> 
> Hi can you please tell me exactly where I have to put this in my TS file?


You can create a function 

   resizeGraph() {
  window.dispatchEvent(new Event('resize'));
  }

and you can call this function on some button which toggles the screen or do some operation related to graph

**Comment by SandeepBalachandran** [comment]
> I solved it by putting it inside a div with css --> display: grid;
> 
> // html
> 
> ```
> <div id="pie-chart-parent">
> 
>     <ngx-charts-pie-chart></ngx-charts-pie-chart>
> 
> </div>
> ```
> 
> // css
> `#pie-chart-parent { display: grid; }`

This one fixed my issue. But i am not quite sure this is the right way .I mean it is way better to tweak these controls from library itself

**Comment by DinalieLiyanage** [comment]
> @dudipsh My onResize function does not get called .I am using [angular-gridster2](https://github.com/tiberiuzuld/angular-gridster2) with ngx-charts so I need to resize the chart based on div size .Here is how I have used
> 
> ```
> <div #component1 class="ngx-charts-bar-vertical" style=" display: grid; ">
>   <ngx-charts-bar-vertical
>   (window:resize)="onResize($event)" 
>   [scheme]="colorScheme"
>   [results]="single"
>   [gradient]="gradient"
>   [xAxis]="showXAxis"
>   [yAxis]="showYAxis"
>   [legend]="showLegend"
>   [showXAxisLabel]="showXAxisLabel"
>   [showYAxisLabel]="showYAxisLabel"
>   [xAxisLabel]="xAxisLabel"
>   [yAxisLabel]="yAxisLabel"
>   [xAxisTickFormatting]="xAxisTickFormatting"
>   (select)="onSelect($event)"
>   (legendLabelClick)="onLegendLabelClick($event)"
>   (dblclick)="dblclick($event)"
>   (xAxisLabelclick)="xAxisLabelclick($event)">
>   
>   </ngx-charts-bar-vertical>
> </div>
> ```
> 
> Any help on how can i resize the chart based on user interaction with div .

This worked for me. Thanks

**Comment by DinalieLiyanage** [comment]
> The solution for me was to create a parent div and add some CSS :
> `charts.component.html`
> 
> ```
> <div id="line-chart-parent" class="col-8">
>  <ngx-charts-line-chart class="chart-container"
>   [scheme]="colorScheme"
>   [results]="results"
>   [curve]="curve"
>   [showXAxisLabel]="true"
>   [showYAxisLabel]="true"
>   [xAxis]="true"
>   [yAxis]="true"
>   [autoScale]="true"
>   [showGridLines]="true"
>   (select)="select($event)">
>  </ngx-charts-line-chart>
> </div>
> ```
> 
> `charts.component.css`
> 
> ```
> #line-chart-parent {
>  display: grid;
>  min-height: 500px;
> }
> ```
> 
> I wanted the width to be responsive and the heigth to be fixed.

This worked for me. Thanks in advance

**Comment by IgorKurkov** [comment]
So `display:grid` doesn't work with advanced-pie, bars. The listeners to rerender chart (width\height) also doesn't work. Will replace these charts with another more responsive, unfortunately.

**Comment by ksohail-weeinvent** [comment]
> Try removing the `[view]` input. If that is not present or undefined, the chart resizes to fit its parent container.

set the parent div width 100%; as well as give some height to that div,this will do the trick.
 <div class="w-100" style="height: 400px;">
                        
                        <ngx-charts-line-chart  [scheme]="colorScheme" [legend]="legend" [showXAxisLabel]="showXAxisLabel" [showYAxisLabel]="showYAxisLabel" [xAxis]="xAxis" [yAxis]="yAxis" [xAxisLabel]="xAxisLabel" [yAxisLabel]="yAxisLabel" [timeline]="timeline" [results]="graphActualContent" (select)="onSelect($event)" (activate)="onActivate($event)" (deactivate)="onDeactivate($event)">
                        </ngx-charts-line-chart>
 </div>

**Comment by dkoshevets** [comment]
I had similar problem, I had grid with two columns with mat-card elements. Inside each mat-card-content I had a chart. When I increased viewport the chart scaled well, but when viewport shrinked the chart didn't change it's size. BUT, after page reload I had desired view. So, i came up to a solution - reload page when page resizes (it's not so often typical user changes the size of viewport, he-he)

`fromEvent(window, 'resize').pipe(
      debounceTime(300),
    ).subscribe((e: Event) => {
      console.log(e);
      window.location.reload();
    });`
